cloud/amazon/ec2_search.py

```diff
@@ -0,0 +1,157 @@
+    required: false
+    default: null
+    aliases: []
+  lookup:
+    description:
+      - What type of lookup to use when searching EC2 instance info.
+    required: false
+    default: tags
+    aliases: []
+  region:
+    description:
+      - EC2 region that it should look for tags in
+    required: false
+    default: All Regions
+    aliases: []
+  ignore_state:
+    description:
+      - instance state that should be ignored such as terminated.
+    required: false
+    default: terminated
+    aliases: []
+author:
+    - "Michael Schuett (@michaeljs1990)"
+extends_documentation_fragment: aws
+'''
+
+EXAMPLES = '''
+# Note: These examples do not set authentication details, see the AWS Guide for details.
+
+# Basic provisioning example
+- ec2_search:
+    key: mykey
+    value: myvalue
+  register: servers
+'''
+try:
+    import boto
+    import boto.ec2
+    HAS_BOTO = True
+except ImportError:
+    HAS_BOTO = False
+
+def todict(obj, classkey=None):
+    if isinstance(obj, dict):
+        data = {}
+        for (k, v) in obj.items():
+            data[k] = todict(v, classkey)
+        return data
+    elif hasattr(obj, "_ast"):
+        return todict(obj._ast())
+    elif hasattr(obj, "__iter__"):
+        return [todict(v, classkey) for v in obj]
+    elif hasattr(obj, "__dict__"):
+        # This Class causes a recursive loop and at this time is not worth
+        # debugging. If it's useful later I'll look into it.
+        if not isinstance(obj, boto.ec2.blockdevicemapping.BlockDeviceType):
+            data = dict([(key, todict(value, classkey))
+                for key, value in obj.__dict__.iteritems()
+                if not callable(value) and not key.startswith('_')])
+            if classkey is not None and hasattr(obj, "__class__"):
+                data[classkey] = obj.__class__.__name__
+            return data
+    else:
+        return obj
+
+def get_all_ec2_regions(module):
+    try:
+        regions = boto.ec2.regions()
+    except Exception, e:
+        module.fail_json('Boto authentication issue: %s' % e)
+
+    return regions
+
+# Connect to ec2 region
+def connect_to_region(region, module):
+    try:
+        conn = boto.ec2.connect_to_region(region.name)
+    except Exception, e:
+        print module.jsonify('error connecting to region: ' + region.name)
+        conn = None
+    # connect_to_region will fail "silently" by returning
+    # None if the region name is wrong or not supported
+    return conn
+
+def main():
+    module = AnsibleModule(
+        argument_spec = dict(
+            key = dict(default='Name'),
+            value = dict(),
+            lookup = dict(default='tags'),
+            ignore_state = dict(default='terminated'),
+            region = dict(),
+        )
+    )
+
+    if not HAS_BOTO:
+        module.fail_json(msg='boto required for this module')
+
+    server_info = list()
+
+    for region in get_all_ec2_regions(module):
+        conn = connect_to_region(region, module)
```

Comment:
I would connect using the inbuilt ansible methods. Something like...

region, ec2_url, aws_connect_params = get_aws_connection_info(module)
    
    if region:
        try:
            conn = connect_to_aws(boto.ec2, region, **aws_connect_params)
        except (boto.exception.NoAuthHandlerFound, StandardError), e:
            module.fail_json(msg=str(e))
    else:
        module.fail_json(msg="region must be specified")